### PR TITLE
systemd-mount features and improvements

### PR DESCRIPTION
**Pull request by lrusak**
I like the systemd-mount tool but it is lacking in some ways, mainly a way to "unmount". I wanted to implement behaviour that is similar to the way `udevil` acts when mounting/unmounting.

In the first commit I basically refactored much of the existing code to allow a command to be specified.

Old behaviour:
`systemd-mount /dev/sdb1`
`systemd-mount --automount=1 /dev/sdb1`
New behaviour:
`systemd-mount mount /dev/sdb1`
`systemd-mount automount /dev/sdb1`

I followed the plumbing of how it is done in the `systemctl` command.

I wanted to do this because I ultimately wanted to provide a "umount" feature which I added in the second commit.
`systemd-mount umount /dev/sdb1`

The first commit is generally quite safe as it was just using the existing code.

The second commit could you a little work I think as I'm not sure if this is the best solution and maybe someone has a better idea of how to achieve this. It also doesn't work for network filesystems at the moment.

Also, it seems that when invoking `systemd-mount umount /dev/sdb1` the mount folder is left behind. I'm not sure what the best way to clean this up would be.

If the first commit is acceptable I can always drop the second commit until a better solution presents itself.

>Disclaimer: I am no C/C++ dev, I am merely a part time hacker who saw something that could be improved

### CODE REVIEW

**Comment by keszybz** [comment]
I'm afraid that such a change is not possible. `systemd-mount` has been released, so the basic calling convention must remain stable.

Also, I'm not sure if the change is warranted. `systemd-mount` is modelled after `mount`, which has separate `mount` and `umount` commands, and requiring a seperate "mount" verb would break this similarity.

That said, the unmounting part looks useful. Any chance you could just implement an `--unmount` flag?

**Comment by lrusak** [comment]
I completely understand. Thanks for writing back.

If you would like to follow compatibility with `mount` and `umount` would it be better to make a seperate `systemd-umount` tool then? Or perhaps a program invocation name where if `systemd-umount` is linked to `systemd-mount` so when it is called it will default to `umount`. I don't feel like a flag `--umount` is appropriate.

**Comment by keszybz** [comment]
> Or perhaps a program invocation name where if systemd-umount is linked to systemd-mount so when it is called it will default to umount

That sounds reasonable. @poettering opinions?

>  I don't feel like a flag --umount is appropriate

Yeah, it does look a bit strange, but that'd be easiest to implement.

**Comment by poettering** [comment]
So, changing the command line interface by introducing verbs goes too far, we need to retain compatibility. I'd be fine with the following additions:

1. `systemd-mount --unmount` (together with an alias `systemd-mount --umount` maybe, since I am sure I myself will run into that), plus a short verson `systemd-mount -u`. 
2. We could install a symlink `systemd-umount` and make systemd-mount recognize if it is started like that, and imply --unmount in that case.

I think combining 1 + 2 makes a ton of sense, but I'd merge even one of them alone.

**Comment by lrusak** [comment]
Updated with the suggested comments.

I didn't want to mess with the makefile so people will have to manually link it (or the distro can do it).
```
ln -s systemd-mount systemd-umount
```

Otherwise it seems to work as intended, but could use more testing
```
~ # systemd-mount /dev/sdb1
Started unit run-media-system-testdisk.automount for mount point: /run/media/system/testdisk
~ # ./systemd-umount /dev/sdb1
Stopped unit run-media-system-testdisk.mount for mount point: /run/media/system/testdisk
Stopped unit run-media-system-testdisk.automount for automount point: /run/media/system/testdisk
```

```
~ # systemd-mount --help
systemd-mount [OPTIONS...] WHAT [WHERE]

Establish a mount or auto-mount point transiently.

  -h --help                       Show this help
...
snip
...
     --list                       List mountable block devices
  -u --umount | --unmount         Unmount a transient mount point
```

**Comment by lrusak** [comment]
pushed a fixup to simplify the way args parsing works

File: src/mount/mount-tool.c [hunk 1]
```diff
@@ -99,6 +100,7 @@ static void help(void) {
                "                                  Set automount unit property\n"
                "     --bind-device                Bind automount unit to device\n"
                "     --list                       List mountable block devices\n"
+               "  -u --umount | --unmount         Unmount a transient mount point\n"
```
Comment (poettering) [inline src/mount/mount-tool.c]: I'd document only one of --umount/--unmount and consider the other one just a friendly internal nicety. Neither the --help text nor the man page should document it...

File: src/mount/mount-tool.c [hunk 2]
```diff
@@ -263,13 +268,21 @@ static int parse_argv(int argc, char *argv[]) {
...
+                        break;
+
                 case '?':
                         return -EINVAL;
 
                 default:
                         assert_not_reached("Unhandled option");
                 }
 
+        if (strstr(program_invocation_short_name, "systemd-umount")) {
```
Comment (poettering) [inline src/mount/mount-tool.c]: Single line if blocks without {} please, see CODING_STYLE. Also I think this bit should be moved before the getopt call loop, so that the command line switches always takr precedence over the implied default derived from argv0

File: src/mount/mount-tool.c [hunk 3]
```diff
@@ -607,6 +620,153 @@ static int start_transient_automount(
         return 0;
 }
 
+static int stop_transient_mount(
+                sd_bus *bus,
```
Comment (poettering) [inline src/mount/mount-tool.c]: This code should actually work for all kinds of mount units not just transient ones. I hence would rename it to just stop_mount()

File: src/mount/mount-tool.c [hunk 4]
```diff
@@ -607,6 +620,153 @@ static int start_transient_automount(
...
+                         ansi_highlight(), mount_unit, ansi_normal(),
+                         ansi_highlight(), arg_mount_where, ansi_normal());
+
+        return 0;
+}
+
+static int stop_transient_automount(
+                sd_bus *bus,
+                char **argv) {
+
```
Comment (poettering) [inline src/mount/mount-tool.c]: The only difference between this call and the stop_mount() call above is really the suffix used for the unit. Hence i think the two tunctions should be one and get an additional suffix parameter.

**Comment by lrusak** [comment]
Thanks for the comments. I have made the suggested changes. Let me know if you see anything else :)

File: src/mount/mount-tool.c [hunk 5]
```diff
@@ -99,6 +100,7 @@ static void help(void) {
                "                                  Set automount unit property\n"
                "     --bind-device                Bind automount unit to device\n"
                "     --list                       List mountable block devices\n"
+               "  -u --umount                     Unmount a transient mount point\n"
```
Comment (keszybz) [inline src/mount/mount-tool.c]: Is it still only "transient" mount points that can be unmounted with this command?

File: src/mount/mount-tool.c [hunk 6]
```diff
@@ -152,7 +156,10 @@ static int parse_argv(int argc, char *argv[]) {
         assert(argc >= 0);
         assert(argv);
 
-        while ((c = getopt_long(argc, argv, "hqH:M:t:o:p:A", options, NULL)) >= 0)
+        if (strstr(program_invocation_short_name, "systemd-umount"))
```
Comment (keszybz) [inline src/mount/mount-tool.c:159]: <strike>`streq`?</strike>

Never mind, `strstr` indeed seems better. For example, under libtool the name will be lt-systemd-umount, and with `streq` this wouldn't work.

File: src/mount/mount-tool.c [hunk 7]
```diff
@@ -152,7 +156,10 @@ static int parse_argv(int argc, char *argv[]) {
...
+                        arg_action = ACTION_UMOUNT;
+
+        while ((c = getopt_long(argc, argv, "hqH:M:t:o:p:A:u", options, NULL)) >= 0)
```
Comment (keszybz) [inline src/mount/mount-tool.c]: Why did you add a colon after `A`? Please remove.

File: src/mount/mount-tool.c [hunk 8]
```diff
@@ -607,6 +618,89 @@ static int start_transient_automount(
...
+
+                r = bus_wait_for_jobs_one(w, object, arg_quiet);
+                if (r < 0)
+                        return r;
+        }
+
+        if (!arg_quiet)
+                log_info("Stopped unit %s%s%s for mount point: %s%s%s",
+                         ansi_highlight(), mount_unit, ansi_normal(),
+                         ansi_highlight(), arg_mount_where, ansi_normal());
```
Comment (keszybz) [inline src/mount/mount-tool.c:682]: That seems duplicate... For mount point units the unit name and the mount point are pretty much the same. Since the mount point is specified as an argument to the program, it seems better to display the unit name here.

File: src/mount/mount-tool.c (hunk 6 above)
Comment (lrusak) [reply src/mount/mount-tool.c:159]: was copy/pasted from here, https://github.com/systemd/systemd/blob/master/src/systemctl/systemctl.c#L7993

File: src/mount/mount-tool.c (hunk 7 above)
Comment (lrusak) [reply src/mount/mount-tool.c]: Sorry, didn't know how getopt worked exactly. Thanks for pointing that out.

>If such a character is followed by a colon, the option requires an argument

https://linux.die.net/man/3/getopt_long

File: src/mount/mount-tool.c (hunk 8 above)
Comment (lrusak) [reply src/mount/mount-tool.c:682]: Sure, I can change it if desired, however it was just a copy/paste from here, https://github.com/systemd/systemd/blob/master/src/mount/mount-tool.c#L488-L490

So maybe if you want to change the message/wording that can be done all together at a later time.

Otherwise for now it looks like:
```
Started unit run-media-system-testdisk.automount for mount point: /run/media/system/testdisk
Stopped unit run-media-system-testdisk.mount for mount point: /run/media/system/testdisk
Stopped unit run-media-system-testdisk.automount for automount point: /run/media/system/testdisk
```

**Comment by lrusak** [comment]
Added another fixup with some small changes. Let me know if we want to change the wording. 

I'd like to just get this in and make a new PR if we want to change the log_info message

File: src/mount/mount-tool.c (hunk 8 above)
Comment (keszybz) [reply src/mount/mount-tool.c:682]: Sure, I wasn't aware of that you're just mirroring the other message. I'm not much of a systemd-mount user so far ;)

File: src/mount/mount-tool.c (hunk 6 above)
Comment (keszybz) [reply src/mount/mount-tool.c:159]: Yeah.

**Comment by lrusak** [comment]
Currently there is one issue and I need some direction on how to fix it.

This currently works for block devices `systemd-umount /dev/sdb1` however when trying to unmount via the path `systemd-umount /run/systemd/system/testdisk` it will fail as it expects a block device. It works if you include another arg, `systemd-umount placeholder /run/systemd/system/testdisk`.

So, in order for this to work on block devices and paths we need to include some kind of check, I'm just not sure exactly the best way to implement this an could use some direction.

**Comment by lrusak** [comment]
This fixes the issue. https://github.com/systemd/systemd/commit/35f85e202e7c101a42c1ded45b0cfcdbaf369d33

I'll do a bit more testing and then I'll PR

**Comment by keszybz** [comment]
Yes, please do. Also see #5227. We *want* multiple arguments to be allowed.
